### ISSUE TITLE
Rephrase assignment-as-condition Best Practice

### ISSUE THREAD

**Issue by erosman** [mdn/content#28257]
### MDN URL

https://developer.mozilla.org/en-US/docs/Web/API/Document/createTreeWalker

### What specific section or headline is this issue about?

Examples

### What information was incorrect, unhelpful, or incomplete?

The example shows:

```js
let currentNode;
while ((currentNode = treeWalker.nextNode())) {
  currentNode.data = currentNode.data.toUpperCase();
}
```

Also ...
```js
while ((currentTextNode = textTreeWalker.nextNode())) {
  currentTextNode.data = encodeURI(currentTextNode.data.replace(/\s+/g, " "));
}
```

### What did you expect to see?

It would be worthwhile to use the best practice in the above example as shown in the following section i.e.
```js
while ((currentNode = treeWalker.nextNode()) !== null) {
  // ...
}
```

### Do you have any supporting links, references, or citations?

> [while: Using an assignment as a condition](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/while#using_an_assignment_as_a_condition)
> 
> But the real best practice is to go a step further and make the code even more clear — by adding a comparison operator to turn the condition into an explicit comparison:
> 
> ```js
> const iterator = document.createNodeIterator(document, NodeFilter.SHOW_COMMENT);
> let currentNode;
> while ((currentNode = iterator.nextNode()) !== null) {
>   console.log(currentNode.textContent.trim());
> }
> ```

**Comment by yarusome** [comment]
@sideshowbarker Out of curiosity: Is there any objective proof of why this is the "real best" practice:

https://github.com/mdn/content/blob/main/files/en-us/web/javascript/reference/statements/while/index.md#L112-L120

The claim of "best" practice is indeed endorsed by linters, though.

**Comment by sideshowbarker** [comment]
> @sideshowbarker Out of curiosity: Is there any objective proof of why this is the "real best" practice:

I actually can’t claim at all to have much real domain experience with what the best practices are for JavaScript. I completely defer to @Josh-Cena on that.

However, as far as I can see now, my rationale for claiming it as a “best practice” in the https://github.com/mdn/content/pull/12774 change was, from that same change, this part:

> Along with preventing any warnings in IDEs and code-linting tools, what that code is actually doing will be much more obvious to anybody coming along later who needs to read and understand it or modify it.

So I guess instead of making a claim that it’s a “best practice” what I could have (and maybe should have) written instead was something like:

>  _But the real way to ensure what that code is actually doing will be much more obvious to anybody coming along later who needs to read and understand it or modify it — as well a to prevent any warnings in IDEs and code-linting tools — is to go a step further add a comparison operator to turn the condition into an explicit comparison:_

And in general, I have come around thinking that in our article content here, we should avoid claiming anything is a “best practice” — because whether or not something is a best practice is inherently subjective. We should instead always explicitly state a specific rationale for why what were a recommending is better than the alternatives.

**Comment by Josh-Cena** [comment]
TBH I was surprised that the reference recommended `while ((currentNode = iterator.nextNode()) !== null)` as "best" because it's less obvious to me than with `!== null` omitted. If readability is readily a concern I would just avoid the assignment-as-condition. This pattern could also work: `while ((currentNode = iterator.nextNode()) && currentNode)`, but for people who understand assignment-as-condition, there's really no need for extra redundancy. I would rephrase that recommendation and update the DOM examples.
